File: 438_find_all_anagrams_in_a_string/findAnagrams.cpp
class Solution {
    public:
        vector<int> findAnagrams(string s, string p) {
            vector<int> res;
    
            int n = s.length();
            int m = p.length();
    
            int freqS[26] = {0};
            int freqP[26] = {0};
    
            if (n < m) {
                return res;
            }
    
            for (int i = 0; i < m; i++) {
                freqS[(int)s[i] - 97]++;
                freqP[(int)p[i] - 97]++;
            }
    
            int start = 0;
            int end = m;
    
            if (std::equal(std::begin(freqS), std::end(freqS), std::begin(freqP))) {
                res.push_back(start);
            }
    
            while (end < n) {
                freqS[(int)s[start] - 97]--;
                freqS[(int)s[end] - 97]++;
    
                if (std::equal(std::begin(freqS), std::end(freqS),
                               std::begin(freqP))) {
                    res.push_back(start + 1);
                }
    
                start++;
                end++;
            }
    
            return res;
        }
    };

File: 438_find_all_anagrams_in_a_string/findAnagrams.go
func findAnagrams(s string, p string) []int {
    n := len(s)
    m := len(p)

    res := make([]int, 0)

    if n < m {
        return res
    }

    freqS := [26]int{}
    freqP := [26]int{}

    for i := 0; i < m; i++ {
        freqS[s[i] - 'a']++
        freqP[p[i] - 'a']++
    }

    start := 0
    end := m

    if freqS == freqP {
        res = append(res, start)
    }

    for end < n {
        freqS[s[start] - 'a']--
        freqS[s[end] - 'a']++

        if freqS == freqP {
            res = append(res, start + 1)
        }

        start++
        end++
    }

    return res
}